src/Microsoft.OData.Core/UriParser/Binders/BinaryOperatorBinder.cs

```diff
@@ -72,6 +72,12 @@ internal static void PromoteOperandTypes(BinaryOperatorKind binaryOperatorKind,
                 throw new ODataException(ODataErrorStrings.MetadataBinder_IncompatibleOperandsError(leftTypeName, rightTypeName, binaryOperatorKind));
             }
 
+            // If comparing an enum with a string, we must convert the string into the enum equivalent
+            if (left.TypeReference.IsEnum() && right.TypeReference.IsString())
+            {
+                right = new ConstantNode(new ODataEnumValue(((ConstantNode)right).Value.ToString(), leftType.Definition.ToString()), ((ConstantNode)right).Value.ToString(), rightType);
+            }
+

```

Comment:
After PromoteOperandTypes sets the type of the string operand to the appropriate enum type, would it be better to put the logic to create the appropriate constant node within  MetadataBindingUtils.ConvertToTypeIfNeeded?

Reply:
Sure. Moved the conversion logic to ConvertToTypeIfNeeded. Looks cleaner this way.